File: deploy-commands.js
/*
**  caminho: deploy-commands.js
**  últimaMod: 16/07/2025 22:28
**  autor: Vico
**  colaboração: ChatGPT, Gemini, Kimi AI
*/

/*
  Script de deploy de slash commands.
  Suporta CLI:
    node deploy-commands.js              -> global
    node deploy-commands.js --guild 123 -> guild 123
  Dependência: npm i minimist
*/

const { REST, Routes } = require('discord.js');
const fs   = require('fs');
const path = require('path');
const args = require('minimist')(process.argv.slice(2));
const config = require('./config.json');

// ----------------------------------------------------------
// Coleta comandos
// ----------------------------------------------------------
const commands = [];
const commandsPath = path.join(__dirname, 'commands');
const commandFiles = fs.readdirSync(commandsPath)
                      .filter(f => f.endsWith('.js') && f !== 'index.js');

for (const file of commandFiles) {
  const cmd = require(path.join(commandsPath, file));
  if ('data' in cmd && 'execute' in cmd) {
    commands.push(cmd.data.toJSON());
  } else {
    console.warn(`[VICA][DEPLOY] Comando ignorado (falta data/execute): ${file}`);
  }
}

const rest = new REST({ version: '10' }).setToken(config.discord.token);

(async () => {
  try {
    const guildIdCLI = args.guild;

    console.log(
      guildIdCLI
        ? `[VICA][DEPLOY] Enviando ${commands.length} comandos para guild ${guildIdCLI}`
        : `[VICA][DEPLOY] Enviando ${commands.length} comandos GLOBALMENTE`
    );

    if (guildIdCLI) {
      await rest.put(
        Routes.applicationGuildCommands(config.discord.clientId, guildIdCLI),
        { body: commands }
      );
      console.log('[VICA][DEPLOY] Comandos de guild enviados com sucesso.');
    } else {
      await rest.put(
        Routes.applicationCommands(config.discord.clientId),
        { body: commands }
      );
      console.log('[VICA][DEPLOY] Comandos globais enviados com sucesso.');
      console.log('[VICA][DEPLOY] Pode levar até 1 hora para aparecer em todos os servidores.');
    }
  } catch (err) {
    console.error('[VICA][DEPLOY] Erro:', err);
  }
})();

File: deploy_commands.py
"""Deploy NeoVica slash commands globally or to one Discord guild."""

from __future__ import annotations

import argparse
import asyncio

from vica.deployment import create_deployment_bot


def discord_id(value: str) -> int:
    try:
        parsed = int(value)
    except ValueError as exc:
        raise argparse.ArgumentTypeError("o ID precisa ser um numero inteiro") from exc
    if parsed <= 0:
        raise argparse.ArgumentTypeError("o ID precisa ser positivo")
    return parsed


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Sincroniza os slash commands da NeoVica com o Discord."
    )
    parser.add_argument(
        "--guild",
        type=discord_id,
        help="ID da guild; sem esta opcao, o deploy sera global.",
    )
    return parser.parse_args()


async def run(guild_id: int | None) -> None:
    bot = await create_deployment_bot()
    try:
        if guild_id is None:
            print("[NEOVICA][DEPLOY] Enviando comandos GLOBALMENTE")
            synced = await bot.sync_commands()
            print(f"[NEOVICA][DEPLOY] {len(synced)} comandos enviados com sucesso.")
            print(
                "[NEOVICA][DEPLOY] Comandos globais podem levar ate 1 hora para aparecer."
            )
        else:
            print(f"[NEOVICA][DEPLOY] Enviando comandos para guild {guild_id}")
            synced = await bot.sync_commands(guild_id)
            print(f"[NEOVICA][DEPLOY] {len(synced)} comandos enviados com sucesso.")
    finally:
        await bot.close()


def main() -> int:
    args = parse_args()
    try:
        asyncio.run(run(args.guild))
    except Exception as exc:
        print(f"[NEOVICA][DEPLOY] Erro: {exc}")
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
